ts/Data/DataTable.ts

```diff
@@ -0,0 +1,902 @@
+     *
+     * @param {DataEventEmitter.EventDetail} [eventDetail]
+     * Custom information for pending events.
+     *
+     * @return {boolean}
+     * Returns true, if the delete was successful, otherwise false.
+     *
+     * @emits DataTable#deleteRow
+     * @emits DataTable#afterDeleteRow
+     */
+    public deleteRow(
+        rowId: string,
+        eventDetail?: DataEventEmitter.EventDetail
+    ): (DataTableRow|undefined) {
+        const row = this.rowsIdMap[rowId],
+            index = this.rows.indexOf(row);
+
+        this.emit({ type: 'deleteRow', detail: eventDetail, index, row });
+
+        this.rows[index] = row;
+        delete this.rowsIdMap[rowId];
+        this.unwatchRow(rowId);
+
+        this.emit({ type: 'afterDeleteRow', detail: eventDetail, index, row });
+
+        return row;
+    }
+
+    /**
+     * Emits an event on this table to all registered callbacks of the given
+     * event.
+     *
+     * @param {DataTable.EventObject} e
+     * Event object with event information.
+     */
+    public emit(e: DataTable.EventObject): void {
+        fireEvent(this, e.type, e);
+    }
+
+    /**
+     * Return the array of row IDs of this table.
+     *
+     * @return {Array<string>}
+     * Array of row IDs in this table.
+     */
+    public getAllRowIds(): Array<string> {
+        return Object.keys(this.rowsIdMap);
+    }
+
+    /**
+     * Returns a copy of the internal array with rows.
+     *
+     * @return {Array<DataTableRow>}
+     * Copy of the internal array with rows.
+     */
+    public getAllRows(): Array<DataTableRow> {
+        return this.rows.slice();
+    }
+
+    /**
+     * Returns the row with the fiven index or row ID.
+     *
+     * @param {number|string} row
+     * Row index or row ID.
+     *
+     * @return {DataTableRow.CellType}
+     * Column value of the column in this row.
+     */
+    public getRow(row: (number|string)): (DataTableRow|undefined) {
+
+        if (typeof row === 'string') {
+            return this.rowsIdMap[row];
+        }
+
+        return this.rows[row];
+    }
+
+    /**
+     * Returns the number of rows in this table.
+     *
+     * @return {number}
+     * Number of rows in this table.
+     *
+     * @todo Consider implementation via property getter `.length` depending on
+     *       browser support.
+     */
+    public getRowCount(): number {
+        return this.rows.length;
+    }
+
+    /**
+     * Returns the version tag that changes with each modification of the table
+     * or a related row.
+     *
+     * @return {string}
+     * Version tag of the current state.
+     */
+    public getVersionTag(): string {
+        return this.versionTag || (this.versionTag = uniqueKey());
+    }
+
+    /**
+     * Converts the datatable instance to a record of columns
+     * @return {DataTable.ColumnCollection}
+     * A record of columns
+     */
+    public toColumns(): DataTable.ColumnCollection {
+        return DataTable.toColumns(this);
+    }
```

Comment:
I find this confusing. We should decide, if this is a static function or instance function. I favor instance like with toJSON and toString, so the implementation of `DataTable.toColumns` would move to this. What do you think?

Reply:
Agree, I mostly made it static to match `fromColumns` but I don't see any use cases where it is necessary.